Improve Regexp validations with early returns and graceful handling of `PatternSyntaxException`

sanjana opened the pull request:
## Purpose
> $title

Partially Fixes #39281
Fixes: #39685
Fixes: #39749

## Approach
> 

## Samples
> 

## Remarks
> 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

pcnfernando commented:
@sanjana Please check on the build failure

LakshanWeerasinghe commented on bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/RuntimeErrors.java:
```diff
@@ -127,7 +127,12 @@ public enum RuntimeErrors implements DiagnosticCode {
     INCOMPATIBLE_ARGUMENTS("incompatible.arguments", "RUNTIME_0101"),
-    DECIMAL_VALUE_OUT_OF_RANGE("decimal.value.out.of.range", "RUNTIME_0102");
+    DECIMAL_VALUE_OUT_OF_RANGE("decimal.value.out.of.range", "RUNTIME_0102"),
+    NEGATIVE_REGEXP_FIND_INDEX("negative.regexp.match.string.index", "RUNTIME_0103"),
```
```suggestion
    NEGATIVE_REGEXP_FIND_INDEX("regexp.match.string.negative.index", "RUNTIME_0103"),
```
In order to have consistency.

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/a7eb91d3a0dd754d2121774779e1e2d7fbe2b64e.

LakshanWeerasinghe commented on bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/RuntimeErrors.java:
```diff
@@ -127,7 +127,12 @@ public enum RuntimeErrors implements DiagnosticCode {
+    NEGATIVE_REGEXP_FIND_INDEX("negative.regexp.match.string.index", "RUNTIME_0103"),
+    INVALID_REGEXP_FIND_INDEX("regexp.match.string.index.out.of.range", "RUNTIME_0104"),
+    EMPTY_STRING("regexp.empty.string", "RUNTIME_0104"),
+    EMPTY_REGEXP("empty.regexp", "RUNTIME_0105"),
```
```suggestion
    EMPTY_REGEXP("regexp.empty", "RUNTIME_0105"),
```

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/e1ae14d4aecebf28ef562a2b13824f852931b87c.

LakshanWeerasinghe commented on bvm/ballerina-runtime/src/main/resources/MessagesBundle.properties:
```diff
@@ -242,3 +242,6 @@ decimal.value.out.of.range = decimal range overflow
 config.size.mismatch = [{0}] the size for configurable variable ''{1}'' is expected to be ''{2}'', \
   but found ''{3}''
 invalid.utf8.byte.array.value = array contains invalid UTF-8 byte value
+negative.regexp.match.string.index = start index cannot be less than 0
```
```suggestion
regexp.match.string.negative.index = start index cannot be less than 0
```

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/a7eb91d3a0dd754d2121774779e1e2d7fbe2b64e.

LakshanWeerasinghe commented on bvm/ballerina-runtime/src/main/resources/MessagesBundle.properties:
```diff
@@ -242,3 +242,6 @@ decimal.value.out.of.range = decimal range overflow
 invalid.utf8.byte.array.value = array contains invalid UTF-8 byte value
+negative.regexp.match.string.index = start index cannot be less than 0
+regexp.match.string.index.out.of.range = start index ''{0}'' cannot be greater than input length ''{1}''
+regexp.invalid.pattern = invalid regexp pattern
```
I think you have missed to add "regexp.empty.string" and "regexp.empty" values.

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/e1ae14d4aecebf28ef562a2b13824f852931b87c.

LakshanWeerasinghe reviewed (changes requested):
Test cases are missing for "RUNTIME_0104" and "RUNTIME_0105".

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:
```diff
@@ -74,9 +94,14 @@ public static BArray findAll(BRegexpValue regExp, BString str, int startIndex) {
+    public static BArray findAllGroups(BRegexpValue regExp, BString str, long startIndex) {
+        checkIndexWithinRange(str, startIndex);
+        if (isEmptyStrOrRegexp(regExp, str)) {
+            return null;
```
Lets cover this

sanjana replied:
Added in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/1f7c4326bad3eda9e064c2de7b125b53e11ce284.

pcnfernando commented:
We have changes in https://github.com/ballerina-platform/ballerina-lang/pull/39238 here as well.

sanjana commented:
> We have changes in #39238 here as well.

Added in https://github.com/ballerina-platform/ballerina-lang/commit/c68dce049bfb41f12e1fdade8ea90598b8bba533.

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java:
```diff
@@ -45,8 +46,12 @@ public class RegexUtil {
+        try {
+            return getMatcher(regexpVal, inputStr.getValue());
+        } catch (Exception e) {
+            throw new PatternSyntaxException("Invalid regex syntax", regexpVal.toString(), -1);
```
Why do you have to wrap every exception inside a PatternSyntaxException?

sanjana replied:
The `PatternSyntaxException` thrown in the method signature is handled here.
https://github.com/ballerina-platform/ballerina-lang/pull/39444/files#diff-5b118eb7fd43169d368376a7e901ef3c2a4d6a12c048d784d1c73c3b14beaf53R49

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/5456a304cee5572dfe0375a7564ce75c6a4d8b11.

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/FromString.java:
```diff
@@ -37,6 +40,11 @@ public class FromString {
+            }
+
             return TypeConverter.stringToRegExp(string.getValue());
         } catch (BError bError) {
```
You should be able to remove this catch

sanjana replied:
Wouldn't removing it cause a compilation error due to a `BError` being thrown at `ypeConverter.stringToRegExp()` for invalid regexp syntax?

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java:
```diff
@@ -53,7 +66,20 @@ public static BArray matchGroupsAt(BRegexpValue regExp, BString str, int startIn
+    private static Matcher getMatcher(BRegexpValue regExp, BString str) {
+        try {
+            return RegexUtil.getMatcher(regExp, str);
+        } catch (Exception e) {
```
Wouldn't catching PatternSyntaxException be enough?

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/4c4adeacdb83b2291c52dff4a09215f5198c305e.

mohanvive commented:
@sanjana Check the comments and resolve the conflicts ASAP.

pcnfernando commented:
@sanjana can you check how/whether the below two differs? Add a test for interpolated case as well
re `` and re `${""}`

sanjana commented:
> @sanjana can you check how/whether the below two differs? Add a test for interpolated case as well re `` and re `${""}`

@pcnfernando will check and add a test case to validate the above case.

sanjana commented:
> @sanjana can you check how/whether the below two differs? Add a test for interpolated case as well re `` and re `${""}`

@pcnfernando added in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/a31ebcf0a39a39786c6786b18148e50e5ab91fe7.

pcnfernando commented:
@sanjana please check on the build failure

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:
```diff
@@ -109,9 +135,14 @@ private static void checkIndexWithinRange(BString str, long startIndex) {
     }
+
+    private static boolean isEmptyRegexp(BRegexpValue regExp) {
+
```
```suggestion
```

sanjana replied:
Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/d2e825dc93401a2bcd7bac4ee797b9304347c3ee.

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java:
```diff
@@ -53,7 +65,20 @@ public static BArray matchGroupsAt(BRegexpValue regExp, BString str, int startIn
+        try {
+            return RegexUtil.getMatcher(regExp, str);
+        } catch (PatternSyntaxException e) {
+            throw BLangExceptionHelper.getRuntimeException(BallerinaErrorReasons.REG_EXP_PARSING_ERROR,
```
Can't we handle this at the util level where we actually create the Pattern?

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/d2e825dc93401a2bcd7bac4ee797b9304347c3ee.

sanjana commented:
> @sanjana please check on the build failure

@pcnfernando Updated the checkstyle failing issue in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/922eef3887bf60b2d6dd70fcf2b2cd558460b45a.

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:
```diff
@@ -109,9 +135,13 @@ private static void checkIndexWithinRange(BString str, long startIndex) {
         }
     }
+
+    private static boolean isEmptyRegexp(BRegexpValue regExp) {
```
Move this to util class and reuse everywhere

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/f4b63bf37c6784813d85b782ccea40b35c7d319c.

SasinduDilshara commented on langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:
```diff
@@ -1060,6 +1060,73 @@ function testLangLibFuncWithNamedArgExpr() {
+
+   //  // find groups
+    regexp:Groups? resC1 = regexp:findGroups(re ``, "Butter was bought by Betty but the butter was bitter");
+    assertTrue(resC1 is null);
```
Don't we need to use `resC1 is ()` instead of `resC1 is null`

sanjana replied:
Will be addressed aligning with the new approach mentioned at: https://github.com/ballerina-platform/ballerina-lang/pull/39444#discussion_r1124231161.

SasinduDilshara commented on langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:
```diff
@@ -1060,6 +1060,73 @@ function testLangLibFuncWithNamedArgExpr() {
+    regexp:Span? resA2 = regexp:matchAt(re = re ``, str = "HelloWorld", startIndex = 4);
+    assertTrue(resA2 is null);
+    string regexStrA = "";
+    regexp:Span? resA3 = regexp:matchAt(re = re `${regexStrA}`, str = "HelloWorld");
```
Can you check this behaviour with Javascript. Imo, This is incorrect since we define our regex pattern as a empty string

sanjana replied:
The string value of ``re `${regexStrA}` `` and ```re `` ``` both will be evaluated to `""`.
@pcnfernando is there any way to distinguish between the above two cases?

sanjana replied:
As per an offline discussion with @mohanvive @pcnfernando @SasinduDilshara it was decided provide a compile-time error for regexps with empty values ``` re `` ```. 
And empty values which are available at run-time, will be mapped to the non-capturing group `(?:)`.

Will be fixing #39794 as well.

SasinduDilshara commented on langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:
```diff
@@ -1060,6 +1060,73 @@ function testLangLibFuncWithNamedArgExpr() {
+    assertTrue(resA7 is null);
+    regexp:Span? resA8 = regexp:find(re = re `(.*)`, str = "");
+    assertTrue(resA8 is regexp:Span);
+
```
Shall we add some tests for `fullMatch` function also.
Add a test with empty regex and a empty string is a better option.

sanjana replied:
Will be addressed aligning with the new approach mentioned at: https://github.com/ballerina-platform/ballerina-lang/pull/39444#discussion_r1124231161.

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:
```diff
@@ -97,6 +115,15 @@ public static BArray findAllGroups(BRegexpValue regExp, BString str, long startI
+        try {
+            return RegexUtil.getMatcher(regExp, str);
+        } catch (PatternSyntaxException e) {
+            throw BLangExceptionHelper.getRuntimeException(BallerinaErrorReasons.REG_EXP_PARSING_ERROR,
```
Let's append `e.message` to the bal error. Check other cases as well

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Find.java:
```diff
@@ -97,6 +115,15 @@ public static BArray findAllGroups(BRegexpValue regExp, BString str, long startI
         return groupArray;
     }
 
+    private static Matcher getMatcher(BRegexpValue regExp, BString str) {
```
Why aren't we calling `RegexUtil.getMatcher()`?

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/6143046ac78ebe122ff65b29772bfdb7f62a2554.

pcnfernando commented on langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/RegexUtil.java:
```diff
@@ -46,7 +50,12 @@ public class RegexUtil {
+        try {
+            return getMatcher(regexpVal, inputStr.getValue());
+        } catch (PatternSyntaxException e) {
+            throw BLangExceptionHelper.getRuntimeException(BallerinaErrorReasons.REG_EXP_PARSING_ERROR,
```
Let's append the `e.message`

sanjana replied:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39444/commits/f9cc62cc815c846f7e99f4711c015f2167bb86eb.

SasinduDilshara commented:
Shall we reomve #39794 from the resolving issue list in this PR
Since our plan is to create a new PR for that

sanjana commented:
> Shall we remove #39794 from the resolving issue list in this PR Since our plan is to create a new PR for that

Removed form the description. Will be addressing it separately. 
Fixed the test failure that occurred in the the windows env due to the newline character issue.

SasinduDilshara commented on langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:
```diff
@@ -1060,6 +1060,73 @@ function testLangLibFuncWithNamedArgExpr() {
+    assertTrue(resA3 is ());
+    regexp:Span? resA4 = regexp:matchAt(re = re `${regexStrA}`, str = "HelloWorld", startIndex = 4);
+    assertTrue(resA4 is ());
+
```
Small suggestion: I think we don't need this new line here. 
Other than that LGTM.

sanjana replied:
This is updated at: #39848.
